### ISSUE TITLE
Activate issue with parallel activations - "File exists" in symlinked path

### ISSUE THREAD

**Issue by tomfleming** [conda/conda#2837]
I'm working in a cluster environment where I am trying to activate a conda environment in multiple parallel processes. After recently upgrading to the latest release (4.1.2), I've been receiving a "File exists" error in some of those processes, similar to the below:

```
Traceback (most recent call last):
  File "/symln_home/anaconda/bin/conda", line 6, in <module>
    sys.exit(main())
  File "/Users/tomflem/anaconda/lib/python2.7/site-packages/conda/cli/main.py", line 48, in main
    activate.main()
  File "/Users/tomflem/anaconda/lib/python2.7/site-packages/conda/cli/activate.py", line 163, in main
    conda.install.symlink_conda(prefix, root_dir, shell)
  File "/Users/tomflem/anaconda/lib/python2.7/site-packages/conda/install.py", line 582, in symlink_conda
    symlink_conda_hlp(prefix, root_dir, where, symlink_fn)
  File "/Users/tomflem/anaconda/lib/python2.7/site-packages/conda/install.py", line 599, in symlink_conda_hlp
    symlink_fn(root_file, prefix_file)
OSError: [Errno 17] File exists
```

I've been able to repro this issue on a laptop when my PATH to the root anaconda install contains a symlinked directory ("symln_home," in this example). Doesn't seem to be a problem when the PATH is symlink-free. The below has produced the error consistently for me on OS X:

```
$ echo $PATH
/symln_home/anaconda/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
```

``` python
from multiprocessing import Pool
import os

call = ['source activate myenv && echo $PATH' for i in range(30)]
pool = Pool(4)
pool.map(os.system, call)
pool.close()
```

Is this a bug, or are there any workarounds for this particular use case? In my cluster environment, my HOME directory (where I've installed anaconda) is symlinked, which I would assume is a fairly common setup in shared computing (I don't admin it, so unfortunately I can't change that aspect).

**Comment by kalefranz** [comment]
@tomfleming We've had some fixes since `4.1.2` that might have tangentially touched on this, and may have ended up helping.  Can you see if you still have the problem as of `4.1.5`?

**Comment by raylim** [comment]
I downgraded to `4.0.0` and it appears to have resolved this issue.

**Comment by mcg1969** [comment]
I'd recommend dropping back not to 4.0.0, but rather 4.0.10.

**Comment by jgmorrison** [comment]
This has been an issue brought to us. When running multiple instances of a bash script in parallel where a conda environment is activated prior to executing some python code, but the error below is produced:

`Traceback (most recent call last): 
File "/usr/prog/anaconda/3.5-4.0.0/bin/conda", line 6, in <module> 
sys.exit(main()) 
File "/usr/prog/anaconda/3.5-4.0.0/lib/python3.5/site-packages/conda/cli/main.py", line 48, in main 
activate.main() 
File "/usr/prog/anaconda/3.5-4.0.0/lib/python3.5/site-packages/conda/cli/activate.py", line 163, in main 
conda.install.symlink_conda(prefix, root_dir, shell) 
File "/usr/prog/anaconda/3.5-4.0.0/lib/python3.5/site-packages/conda/install.py", line 582, in symlink_conda 
symlink_conda_hlp(prefix, root_dir, where, symlink_fn) 
File "/usr/prog/anaconda/3.5-4.0.0/lib/python3.5/site-packages/conda/install.py", line 599, in symlink_conda_hlp 
symlink_fn(root_file, prefix_file) 
FileExistsError: [Errno 17] File exists: '/usr/prog/anaconda/3.5-4.0.0/bin/conda' -> '/home/john/.conda/envs/scikit/bin/conda'`

The script works as expected when it's not executed concurrently. The error was worked around by replacing the activation of the environment with the corresponding $PATH settings.

**Comment by kalefranz** [comment]
#3210 should patch over the issue, but https://github.com/conda/conda/issues/2502 is probably the "correct" fix.

**Comment by kalefranz** [comment]
Fix will be included in 4.1.12.

**Comment by leepc12** [comment]
This bug still persists even in the latest version (4.2.1).

**Comment by dandaman** [comment]
A highly related issue occurs if two parallel jobs source activate at the same time in row 609 of install.py:
```
if os.path.lexists(prefix_file):
                os.remove(prefix_file)
```

the process that is "slower" in removing the symlink, dies because there is no specific handling of errno.ENOENT.

In my application I can overcome this by adding an elif for errno.ENOENT:

```
try:
            # try to kill stale links if they exist
            if os.path.lexists(prefix_file):
                os.remove(prefix_file)
            # if they're in use, they won't be killed.  Skip making new symlink.
            if not os.path.lexists(prefix_file):
                symlink_fn(root_file, prefix_file)
        except (IOError, OSError) as e:
            if (os.path.lexists(prefix_file) and
                    (e.errno in (errno.EPERM, errno.EACCES, errno.EROFS, errno.EEXIST))):
                log.debug("Cannot symlink {0} to {1}. Ignoring since link already exists.".format(root_file, prefix_file))
            elif (e.errno in (errno.ENOENT)):
                log.debug("Problem with symlink management {0} {1}. File might have been removed by another process in the meantime." .format(root_file, prefix_file))
            else:
                raise

```

Not sure whether this is a generally applicable fix. Better would be a solution with independent handling of parallel processes using conda environments...Maybe by creating/using randomized directories for each source activate....

**Comment by yaroslavvb** [comment]
I'm running into the issue with latest conda (4.2.0), @raylim how did you downgrade to 4.0?

**Comment by croth1** [comment]
Ran into the same problem with conda=4.2.13.

**Comment by raylim** [comment]
Yes @yaroslavvb, I'm running conda=4.0.10

**Comment by croth1** [comment]
@kalefranz, any progress or new issue for tracking this problem? Currently it's really painful to run conda from a shared file system on a cluster.

**Comment by geoffreyphipps-uw** [comment]
Ran into the same problem in 4.2.10. Interestingly it only occurred with the zmq package:




    Traceback (most recent call last):
      File "/homes/gphipps/miniconda2/lib/python2.7/site-packages/conda/exceptions.py", line 473, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/homes/gphipps/miniconda2/lib/python2.7/site-packages/conda/cli/main.py", line 93, in _main
        activate.main()
      File "/homes/gphipps/miniconda2/lib/python2.7/site-packages/conda/cli/activate.py", line 148, in main
        conda.install.symlink_conda(prefix, context.root_dir, shell)
      File "/homes/gphipps/miniconda2/lib/python2.7/site-packages/conda/install.py", line 532, in symlink_conda
        symlink_conda_hlp(prefix, root_dir, where, symlink_fn)
      File "/homes/gphipps/miniconda2/lib/python2.7/site-packages/conda/install.py", line 549, in symlink_conda_hlp
        symlink_fn(root_file, prefix_file)
    OSError: [Errno 17] File exists

Traceback (most recent call last):
  File "/snfs2/HOME/gphipps/dalynator_python/dalynator/remote_run_pipeline.py", line 2, in <module>
    from jobmon.job import SGEJob
  File "/homes/gphipps/miniconda2/lib/python2.7/site-packages/jobmon-0.1.5.dev1+ng3d551a7.d20161121-py2.7.egg/jobmon/job.py", line 6, in <module>
    from jobmon.sender import Sender
  File "/homes/gphipps/miniconda2/lib/python2.7/site-packages/jobmon-0.1.5.dev1+ng3d551a7.d20161121-py2.7.egg/jobmon/sender.py", line 7, in <module>
    import zmq
ImportError: No module named zmq

**Comment by geoffreyphipps-uw** [comment]
I think the bug is here, on line 549  of ~/miniconda2/lib/python2.7/site-packages/conda/install.py

if not os.path.lexists(prefix_file):
                symlink_fn(root_file, prefix_file)

If two or more processes are attempting to symlnk the same file then there is a race - both might attempt to create the symlink and the second one will throw an exception.

**Comment by dansondergaard** [comment]
I just experienced this issue with version 4.2.13 in a cluster environment:

```
An unexpected error has occurred.
Please consider posting the following information to the
conda GitHub issue tracker at:

    https://github.com/conda/conda/issues



Current conda install:

               platform : linux-64
          conda version : 4.2.13
       conda is private : False
      conda-env version : 4.2.13
    conda-build version : 2.0.10
         python version : 2.7.12.final.0
       requests version : 2.9.1
       root environment : /home/das/anaconda  (writable)
    default environment : /home/das/anaconda
       envs directories : /home/das/anaconda/envs
          package cache : /home/das/anaconda/pkgs
           channel URLs : https://conda.anaconda.org/bioconda/linux-64
                          https://conda.anaconda.org/bioconda/noarch
                          https://conda.anaconda.org/r/linux-64
                          https://conda.anaconda.org/r/noarch
                          file:///faststorage/data/conda-mirrors/r/linux-64
                          file:///faststorage/data/conda-mirrors/r/noarch
                          file:///faststorage/data/conda-mirrors/bioconda/linux-64
                          file:///faststorage/data/conda-mirrors/bioconda/noarch
                          https://repo.continuum.io/pkgs/free/linux-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/pro/linux-64
                          https://repo.continuum.io/pkgs/pro/noarch
            config file : /home/das/.condarc
           offline mode : False



`$ /home/das/anaconda/bin/conda ..checkenv bash cup`




    Traceback (most recent call last):
      File "/home/das/anaconda/lib/python2.7/site-packages/conda/exceptions.py", line 479, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/home/das/anaconda/lib/python2.7/site-packages/conda/cli/main.py", line 94, in _main
        activate.main()
      File "/home/das/anaconda/lib/python2.7/site-packages/conda/cli/activate.py", line 148, in main
        conda.install.symlink_conda(prefix, context.root_dir, shell)
      File "/home/das/anaconda/lib/python2.7/site-packages/conda/install.py", line 541, in symlink_conda
        symlink_conda_hlp(prefix, root_dir, where, symlink_fn)
      File "/home/das/anaconda/lib/python2.7/site-packages/conda/install.py", line 558, in symlink_conda_hlp
        symlink_fn(root_file, prefix_file)
    OSError: [Errno 17] File exists
```

**Comment by alimanfoo** [comment]
FWIW I'm getting this too with conda 4.2.13 used in a cluster, definitely a race condition as some jobs succeed but others hit this error.

```
An unexpected error has occurred.
Please consider posting the following information to the
conda GitHub issue tracker at:

    https://github.com/conda/conda/issues



Current conda install:

               platform : linux-64
          conda version : 4.2.13
       conda is private : False
      conda-env version : 4.2.13
    conda-build version : not installed
         python version : 3.5.2.final.0
       requests version : 2.10.0
       root environment : /home/aliman/miniconda3  (writable)
    default environment : /home/aliman/miniconda3
       envs directories : /home/aliman/miniconda3/envs
          package cache : /home/aliman/miniconda3/pkgs
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://conda.anaconda.org/bioconda/linux-64
                          https://conda.anaconda.org/bioconda/noarch
                          https://conda.anaconda.org/r/linux-64
                          https://conda.anaconda.org/r/noarch
                          https://repo.continuum.io/pkgs/free/linux-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/pro/linux-64
                          https://repo.continuum.io/pkgs/pro/noarch
            config file : /home/aliman/.condarc
           offline mode : False



`$ /home/aliman/miniconda3/bin/conda ..checkenv bash biipy`




    Traceback (most recent call last):
      File "/home/aliman/miniconda3/lib/python3.5/site-packages/conda/exceptions.py", line 479, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/home/aliman/miniconda3/lib/python3.5/site-packages/conda/cli/main.py", line 94, in _main
        activate.main()
      File "/home/aliman/miniconda3/lib/python3.5/site-packages/conda/cli/activate.py", line 148, in main
        conda.install.symlink_conda(prefix, context.root_dir, shell)
      File "/home/aliman/miniconda3/lib/python3.5/site-packages/conda/install.py", line 541, in symlink_conda
        symlink_conda_hlp(prefix, root_dir, where, symlink_fn)
      File "/home/aliman/miniconda3/lib/python3.5/site-packages/conda/install.py", line 558, in symlink_conda_hlp
        symlink_fn(root_file, prefix_file)
    FileExistsError: [Errno 17] File exists: '/home/aliman/miniconda3/bin/conda' -> '/home/aliman/miniconda3/envs/biipy/bin/conda'
```

Should be re-opened IMO.

**Comment by kalefranz** [comment]
We actually have a fix for this now coming out in 4.3.6 in 24 hours or so. If you want to test, it'll be posted to the conda-canary channel in about 30 minutes or less. 

    conda config --add channels conda-canary
    conda update conda

**Comment by raylim** [comment]
I still get the occasional error where it fails to find `conda`:

```activate: line 55: conda: No such file or directory```

**Comment by raylim** [comment]
That error appears to happen with some regularity.  Anyone else experiencing it?  It's probably some sort of race condition again.

**Comment by raylim** [comment]
And now I can't roll back
```
$ conda install conda=4.0.10
Fetching package metadata .....................
Solving package specifications: .

Package plan for installation in environment /home/limr/usr/anaconda:

The following packages will be SUPERCEDED by a higher-priority channel:

    conda:     4.3.8-py27_0  anaconda    --> 4.0.10-py27_0
    conda-env: 2.6.0-0       anaconda    --> 2.4.5-py27_0

The following packages will be DOWNGRADED due to dependency conflicts:

    requests:  2.12.4-py27_0 conda-forge --> 2.11.1-py27_0 conda-forge

Proceed ([y]/n)? y

ERROR conda.core.link:_execute_actions(318): An error occurred while uninstalling package 'anaconda::conda-4.3.8-py27_0'.
OSError(13, 'Permission denied')
```

Might have to reinstall anaconda.

**Comment by geoffreyphipps-uw** [comment]
Yes, we still experience the bug. We put in a retry-loop as a workaround – so far the retry loop is working. Here:

export PATH=$1:$PATH
rc=17  # anything but zero.
tries=0
max_tries=3
while [[ $rc -ne 0 && $tries -lt $max_tries ]]; do
  source activate $2
  rc=$?
  tries=`expr $tries + 1`
  if [ $tries -gt 1 ] ; then
     # Use an ethernet-style increasing backoff
     jitter=`expr $RANDOM % 10 + 1`
     sleep `expr $jitter \* $tries`
     >&2 echo "jobmon.env_submit_master: WARNING source activate failed, retrying, attempt: $tries"
  fi
done

if [ $tries -ge $max_tries ] ; then
  >&2 echo "jobmon.env_submit_master: ERROR source activate failed"
  exit
fi

# get rid of the env name
shift 2
"$@"

Cheers
Geoff

**Comment by raylim** [comment]
ok thanks, except I think you have to reset the PATH if sourcing the anaconda environment fails.

**Comment by jharenza** [comment]
Any updates on this? Below is the same error and above, there looked to be a fix, but no one confirmed, so I don't want to update without confirmation of the fix.

Thanks!

`An unexpected error has occurred.
Please consider posting the following information to the
conda GitHub issue tracker at:

    https://github.com/conda/conda/issues



Current conda install:

               platform : linux-64
          conda version : 4.2.12
       conda is private : False
      conda-env version : 4.2.12
    conda-build version : not installed
         python version : 3.5.2.final.0
       requests version : 2.11.1
       root environment : /home/harenzaj/miniconda3  (writable)
    default environment : /home/harenzaj/miniconda3
       envs directories : /home/harenzaj/miniconda3/envs
          package cache : /home/harenzaj/miniconda3/pkgs
           channel URLs : https://repo.continuum.io/pkgs/free/linux-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/pro/linux-64
                          https://repo.continuum.io/pkgs/pro/noarch
            config file : None
           offline mode : False



`$ /home/harenzaj/miniconda3/bin/conda ..checkenv bash py27`




    Traceback (most recent call last):
      File "/home/harenzaj/miniconda3/lib/python3.5/site-packages/conda/exceptions.py", line 479, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/home/harenzaj/miniconda3/lib/python3.5/site-packages/conda/cli/main.py", line 94, in _main
        activate.main()
      File "/home/harenzaj/miniconda3/lib/python3.5/site-packages/conda/cli/activate.py", line 148, in main
        conda.install.symlink_conda(prefix, context.root_dir, shell)
      File "/home/harenzaj/miniconda3/lib/python3.5/site-packages/conda/install.py", line 536, in symlink_conda
        symlink_conda_hlp(prefix, root_dir, where, symlink_fn)
      File "/home/harenzaj/miniconda3/lib/python3.5/site-packages/conda/install.py", line 553, in symlink_conda_hlp
        symlink_fn(root_file, prefix_file)
    FileExistsError: [Errno 17] File exists: '/home/harenzaj/miniconda3/bin/deactivate' -> '/home/harenzaj/miniconda3/envs/py27/bin/deactivate'

`

**Comment by raylim** [comment]
Doesn't seem to be fixed. I rolled back to `4.0.10`

**Comment by jharenza** [comment]
When I source activate once in my script before the sample-specific commands I am tasking, the error goes away, rather than activating per sample.

`system("source activate py27");
system("package.py command $sample ... ");`

vs 

`system("source activate py27 && package.py command $sample ...");`

**Comment by holtgrewe** [comment]
Could you reopen the issue? I'm getting precisely the same error with `4.2.13`.

**Comment by kalefranz** [comment]
The latest conda versions are 4.2.16 and 4.3.17. If you can confirm there's still an issue with either of those, then we'll reopen.

**Comment by raylim** [comment]
I can confirm that this is still an issue with the latest conda

**Comment by kalefranz** [comment]
The root cause for this is resolved with #5044 and follow-on PRs in conda 4.4, which will be rolling through the canary process over the next month.
